toolbox/src/base/RobotInterface.cpp

```diff
@@ -0,0 +1,258 @@
+{
+    return getDevice(m_yarpDevices.iPos);
+}
+
+const std::weak_ptr<yarp::dev::IPositionDirect> RobotInterface::getIPositionDirect()
+{
+    return getDevice(m_yarpDevices.iPosDirect);
+}
+
+const std::weak_ptr<yarp::dev::IVelocityControl> RobotInterface::getIVelocityControl()
+{
+    return getDevice(m_yarpDevices.iVel);
+}
+
+const std::weak_ptr<yarp::dev::ITorqueControl> RobotInterface::getITorqueControl()
+{
+    return getDevice(m_yarpDevices.iTorque);
+}
+
+const std::weak_ptr<yarp::dev::IPWMControl> RobotInterface::getIPWMControl()
+{
+    return getDevice(m_yarpDevices.iPWM);
+}
+
+const std::weak_ptr<yarp::dev::ICurrentControl> RobotInterface::getICurrentControl()
+{
+    return getDevice(m_yarpDevices.iCurrent);
+}
+
+const std::weak_ptr<yarp::dev::IEncoders> RobotInterface::getIEncoders()
+{
+    return getDevice(m_yarpDevices.iEncoders);
+}
+
+const std::weak_ptr<yarp::dev::IControlLimits2> RobotInterface::getIControlLimits2()
+{
+    return getDevice(m_yarpDevices.iLimits);
+}
+
+const std::weak_ptr<yarp::dev::IPidControl> RobotInterface::getIPidControl()
+{
+    return getDevice(m_yarpDevices.iPidControl);
+}
+
+// LAZY EVALUATION
+// ===============
+
+bool RobotInterface::retainRemoteControlBoardRemapper()
+{
+    if (m_robotDeviceCounter > 0) {
+        m_robotDeviceCounter++;
+        return true;
+    }
+
+    assert(!m_robotDevice);
+    if (m_robotDevice) {
+        m_robotDevice.reset();
+    }
+
+    if (!initializeRemoteControlBoardRemapper()) {
+        return false;
+    }
+
+    m_robotDeviceCounter++;
+    return true;
+}
+
+bool RobotInterface::releaseRemoteControlBoardRemapper()
+{
+    // The RemoteControlBoardRemapper has not been used
+    if (m_robotDeviceCounter == 0) {
+        return true;
+    }
+
+    // If there are at most 2 blocks with th CB still used
+    if (m_robotDeviceCounter > 1) {
+        m_robotDeviceCounter--;
+        return true;
+    }
+
+    // This should be executed by the last block which uses CB (m_robotDeviceCounter=1)
+    assert(m_robotDevice);
+    if (m_robotDevice) {
+        //  Close the device
+        m_robotDevice->close();
+        // Free the object
+        m_robotDevice.reset();
+        // Free all the drivers
+        m_yarpDevices.iControlMode.reset();
+        m_yarpDevices.iPos.reset();
+        m_yarpDevices.iPosDirect.reset();
+        m_yarpDevices.iVel.reset();
+        m_yarpDevices.iTorque.reset();
+        m_yarpDevices.iPWM.reset();
+        m_yarpDevices.iCurrent.reset();
+        m_yarpDevices.iEncoders.reset();
+        m_yarpDevices.iLimits.reset();
+        m_yarpDevices.iPidControl.reset();
+    }
+
+    m_robotDeviceCounter = 0;
+    return true;
+}
+
+// INITIALIZATION HELPERS
+// ======================
+
+bool RobotInterface::initializeModel()
+{
+    assert (!m_kinDynComp);
+
+    // Allocate the object
+    m_kinDynComp = std::make_shared<iDynTree::KinDynComputations>();
```

Comment:
I found this https://stackoverflow.com/questions/20895648/difference-in-make-shared-and-normal-shared-ptr-in-c

Reply:
Great!
I am a bit scared of
```
Since there there's only one allocation, the pointee's memory cannot be deallocated until the control block is no longer in use. A weak_ptr can keep the control block alive indefinitely.
```
but I think I have to improve a bit my knowledge of shared ptrs before understanding correctly this

Reply:
By the end of the story, it's just the container that stores the counter. In any case its memory footprint would be almost 0.